Fee smearing

pbutti opened the pull request.

pbutti committed:
Fix getP to return const

pbutti committed:
Add smearing function

pbutti committed:
remove mistaken submodule submission

pbutti committed:
Closure test smearing

pbutti committed:
Fixes to smearing tool

pbutti committed:
Added smearing files, config and README

cbravo135 commented on event/include/Track.h:
```diff
@@ -288,9 +288,9 @@ class Track : public TObject {
+        double getP() const {return sqrt(px_*px_ + py_*py_ + pz_*pz_);};
         
-        double getPt() {return sqrt(px_*px_ + pz_*pz_);}
+        double getPt() const {return sqrt(px_*px_ + pz_*pz_);}
```
This should be sqrt(px_*px_ + py_*py_) as the x-y plane is the transverse direction.

cbravo135 commented on processors/include/TrackingAnaProcessor.h:
```diff
@@ -113,11 +115,24 @@ class TrackingAnaProcessor : public Processor {
         std::shared_ptr<TrackSmearingTool> smearingTool_;
+        std::shared_ptr<TrackSmearingTool> smearingToolRel_;
+        std::string pSmearingFile_{""};
+        TH1D* psmear_h_;
```
should we add these as plots in the TrackingHistos instead?

tomeichlersmith replied:
I don't think so. They are explicitly helpful only for those interested in studying the difference smearing can cause. Everyone else would just like smearing to be applied (or not).

cbravo135 reviewed (commented):
Looks like you need to pull the current master as you have some conflicts that need to be resolved

cbravo135 commented:
@alspellm also reported that this branch breaks his analysis code, so we should be sure it doesn't before merging this

alspellm commented:
That's not the case. Was a miscommunication on my part.

On Mon, Feb 12, 2024, 5:10 PM Cameron Bravo ***@***.***>
wrote:

> @alspellm <https://github.com/alspellm> also reported that this branch
> breaks his analysis code, so we should be sure it doesn't before merging
> this
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JeffersonLab/hpstr/pull/187#issuecomment-1940015112>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALKJUATLTIJDU5L77X7DHLTYTK4RXAVCNFSM6AAAAABC2K34BSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNBQGAYTKMJRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tomeichlersmith committed:
Merge branch 'master' into feeSmearing

Conflict resolution
- Undo deleition of TrackSmearingTool header/source
- keep changes made to TrackingAnaProcessor header

tomeichlersmith committed:
get compiling post merge

tomeichlersmith committed:
have pt calculated with px and py

tomeichlersmith committed:
fix up markdown syntax

tomeichlersmith reviewed (commented):
I haven't test run this, but I have confirmed it compiles within my environment.

tomeichlersmith committed:
add a function to apply smearing to a track

tomeichlersmith committed:
need to put utils after event

tomeichlersmith committed:
add InvM histogram calculated from ele and pos directly

introduce TrackSmearingTool into NewVertexAnaProcessor in same way it is
used in TrackingAnaProcessor, compiles and runs without a smearing file
provided (i.e. no effect as intended)

tomeichlersmith committed:
add smearing file to analysis config

this triggers the smearing to be attempted, but it DOES NOT WORK, giving
the following error when attempting to run

```
Error:Bin not found in smearing histogram
---- [ hpstr ]: Event processing complete  --------
...TCLass::LoadClassInfo errors...
```

tomeichlersmith commented:
So I ran into this error when attempting to run. I enabled `debug_` in `TrackSmearingTool` locally in order to get the extra printouts.

```
Track nhits=12 bin=3
Track isTop: 1 nHits: 12 p: 0.993493 deltaP=-0.16636 p'=0.827133
Track nhits=12 bin=3
Track isTop: 0 nHits: 12 p: 0.834609 deltaP=-0.0315946 p'=0.803014
Track nhits=12 bin=3
Track isTop: 0 nHits: 12 p: 0.698569 deltaP=0.105552 p'=0.804121
Track nhits=11 bin=2
Track isTop: 1 nHits: 11 p: 0.896728 deltaP=-0.0805587 p'=0.816169
Track nhits=11 bin=2
Track isTop: 0 nHits: 11 p: 0.622917 deltaP=-0.0144043 p'=0.608513
Track nhits=10 bin=1
Track isTop: 1 nHits: 10 p: 1.25236 deltaP=-0.163196 p'=1.08916
Track nhits=12 bin=3
Track isTop: 1 nHits: 12 p: 0.805661 deltaP=-0.0290028 p'=0.776658
Track nhits=12 bin=3
Track isTop: 0 nHits: 12 p: 1.06335 deltaP=-0.00297324 p'=1.06038
Track nhits=10 bin=1
Track isTop: 0 nHits: 10 p: 0.902764 deltaP=-0.103399 p'=0.799365
Track nhits=12 bin=3
Track isTop: 1 nHits: 12 p: 1.03264 deltaP=0.123815 p'=1.15646
Track nhits=10 bin=1
Track isTop: 1 nHits: 10 p: 0.910685 deltaP=0.120995 p'=1.03168
Track nhits=12 bin=3
Track isTop: 0 nHits: 12 p: 1.12335 deltaP=-0.13851 p'=0.984837
Track nhits=9 bin=0
Error:Bin not found in smearing histogram
---- [ hpstr ]: Event processing complete  --------
```

It looks like there isn't a bin for 9-hit tracks which makes sense if the smearing histogram was created from FEEs. What should I do with these tracks? Smear with the 10-hit smearing? Skip smearing? Cut them out?

tomeichlersmith commented:
Ok, I forced a test by just using the lowest hit count bin in the available smearing histogram.

```diff
@@ -43,7 +42,11 @@ double TrackSmearingTool::smearTrackP(const Track& track) {
   if (debug_)
     std::cout<<"Track nhits="<<nhits<<" bin="<<binN<<std::endl;
   
-  if (binN < 1 || binN > smearing_histo_top_->GetXaxis()->GetNbins()) {
+  if (binN < 1) {
+    std::cout<<"Track nhits="<<nhits<<" bin="<<binN;
+    binN=1;
+    std::cout<<" rounding to bin=1"<< std::endl;
+  } else if (binN > smearing_histo_top_->GetXaxis()->GetNbins()) {
     throw std::invalid_argument("Bin not found in smearing histogram");
   }
```

This seemed to allow for smearing so I moved to checking if the smearing was applied. I am using the 30MeV mass point A' tuples as a test point because the file is quick to run over.[^1] First, looking at the Psum, I am hopeful. We see some broadening due to the smearing applied but the overall shape is maintained.

![ap-30-psum-smear](https://github.com/JeffersonLab/hpstr/assets/31970302/cf1c92f7-ddb8-439d-85dd-53a5f9ef4982)

However, when I try to look at Vertex-level variables recalculating them with the smeared four-momenta, I get a very different result. Below, I compare the smeared four-momenta mass (`(p_ele+p_pos).M()`) against the mass read in from the input file with the reconstructed vertex (`vtx->getInvM()`). These are wildly different. I printed out the value of the manually calculated mass and it just appears to be really wrong most of the time. This would make sense if the Vertex fit is getting the InvM through a smarter calculation than just naively combining the track momentum and cluster energy like I'm doing.

![ap-30-smear-comp](https://github.com/JeffersonLab/hpstr/assets/31970302/2b3e3afe-8617-46df-8781-b64a993c46dd)

[^1]: Available at JLab: `/w/hallb-scshelf2102/hps/teichler/re-reach-vanilla/tuples/ap-30/ap-30_hpstr_tuples.root`

tomeichlersmith committed:
get smearing operational (kinda)

The smearing is now operational in the following sense
- It is applied to any tracks used for later filling, so their momentum manitudes are smeared according to the number of hits
- It is manually applied to the invariant mass using the Moeller smearing factor sqrt(p1_smear/p1_unsmear * p2_smear/p2_unsmear)
- The smearing is NOT applied to anything else
- Any tracks with nhits below the minimum is smeared with the width of the minimum in the smearing histogram and a message is printed

tomeichlersmith commented:
Alright, following @cbravo135 's advice, I am applying the Moeller invariant mass smearing factor as calculated from the tracks that eventually formed the vertex. For validation purposes, I am creating a separated smeared histogram so we can see the smeared (`vtx_smear_InvM_h`) compared to the unsmeared (`vtx_InvM_h`).

![ap-30-moeller-smearing](https://github.com/JeffersonLab/hpstr/assets/31970302/295ed1bf-6347-44b2-b2d9-183d8e033323)

tomeichlersmith committed:
use relative smearing instead of absolute

we want relative smearing since the momenta of these tracks is not
expected to be near 2.3GeV like the FEE tracks used to estimate the
smearing widths

tomeichlersmith commented:
using relative smearing gives more appropriate results. In addition, when looking at invariant mass, we should use the `radMatchTight` region since that requires the **rad**iated electron to be tightly matched to truth MC particles.

![ap-30-moeller-smearing(2)](https://github.com/JeffersonLab/hpstr/assets/31970302/f7c46994-c294-438a-82c5-a485e01f725c)

tomeichlersmith commented:
More thorough (and direct) checking of the momentum smearing can be done by applying the smearing to a known distribution of momenta rather than using a more complicated simulation and reconstruction to provide the testing distribution. The list below was written by @pbutti as some ideas on what to test (of varying levels of complexity).

- Smear fixed momentum and check if you get a gaussian
- Smear a gaussian and check that you get a gaussian
- Generate a mass by choosing 2 fixed momenta p1 and p2. Smear the momenta and check the smeared mass distribution
- Generate 2 momenta p1 and p2 with gaussian distributions, smear them, check the mass distribution.
- Generate 2 momenta p1 and p2 with non gaussian distributions (like throw from a crystal ball with large low tail). Smear those and check the mass distribution.

tomeichlersmith committed:
put notice of bin rounding in debug

tomeichlersmith committed:
add test exectuble for viewing known momentum distortions

tomeichlersmith commented:
Alright, I've added a simple executable that can do trials directly testing the smearing tool itself. Right now, it just does the first two of tests listed above, but that easily confirms that the tool is working as expected. With this evidence, along with the update from Tuesday showing that the correct selection brings the final mass resolution closer to the expected result, I think this is ready to merge.

@cbravo135 Can you re-review?

![direct-smearing-test-fixed-input](https://github.com/JeffersonLab/hpstr/assets/31970302/8d72bcdd-6c47-4744-8843-651ad540b233)
![direct-smearing-test-gaussian-input](https://github.com/JeffersonLab/hpstr/assets/31970302/c7eccea7-1886-4f5e-8b7b-926bc82f3e4c)

pbutti commented:
Hi @cbravo135 @tomeichlersmith 
Anything preventing this branch to be merged into master? 
I would like to add some track biasing support for this tool as well (e.g. p -> p + deltaP). This could be used to improve plotting, momentum estimate and also take into account track-level systematics in analysis. But before making additional changes I would like the current version to go in hpstr.

tomeichlersmith reviewed (approved):
Nothing on my side! For me its all good to go :)
